tests/test_serialize.py

```diff
@@ -1115,3 +1115,48 @@ class MyModel(BaseModel):
         bar: MyEnum
 
     assert MyModel(foo={MyEnum.A: 'hello'}, bar=MyEnum.B).model_dump_json() == '{"foo":{"a":"hello"},"bar":"b"}'
+
+
+@pytest.mark.xfail(reason='Needs new release of pydantic-core')
+def test_subclass_support_unions() -> None:
+    class Pet(BaseModel):
+        name: str
+
+    class Dog(Pet):
+        breed: str
+
+    class Kid(BaseModel):
+        age: str
+
+    class Home(BaseModel):
+        little_guys: Union[List[Pet], List[Kid]]
+
+    class Shelter(BaseModel):
+        pets: List[Pet]
+
+    h1 = Home(little_guys=[Pet(name='spot')])
+    assert h1.model_dump() == {'little_guys': [{'name': 'spot'}]}
```

Comment:
It seems useful to have at a least a couple of items in each list just to ensure that there is no difference between first and other elements.